src/simple.jl

```diff
@@ -1,96 +1,115 @@
+    A::matT
+    x::vecT
+    tol::numT
+    maxiter::Int
+    θ::eigvalT
+    r::vecT
+    Ax::vecT
+    residual::numT
+end
 
-function powm(A;
-    x=nothing, tol::Real=eps(real(eltype(A)))*size(A,2)^3, maxiter::Int=size(A,2),
-    log::Bool=false, kwargs...
-    )
-    K = KrylovSubspace(A, 1)
-    x==nothing ? initrand!(K) : init!(K, x/norm(x))
 
-    history = ConvergenceHistory(partial=!log)
-    history[:tol] = tol
-    reserve!(history,:resnorm, maxiter)
-    eig, v = powm_method!(history, K; tol=tol, maxiter=maxiter, kwargs...)
-    log && shrink!(history)
-    log ? (eig, v, history) : (eig, v)
+##
+## Iterators
+##
+
+@inline converged(p::PowerMethodIterable) = p.residual ≤ p.tol
+
+@inline start(p::PowerMethodIterable) = 0
+
+@inline done(p::PowerMethodIterable, iteration::Int) = iteration > p.maxiter || converged(p)
+
+function next(p::PowerMethodIterable, iteration::Int)
+
+    A_mul_B!(p.Ax, p.A, p.x)
+
+    # Rayleigh quotient θ = x'Ax
+    p.θ = dot(p.x, p.Ax)
+
+    # (Previous) residual vector r = Ax - λx
+    copy!(p.r, p.Ax)
+    BLAS.axpy!(-p.θ, p.x, p.r)
+
+    # Normed residual
+    p.residual = norm(p.r)
+
+    # Normalize the next approximation
+    copy!(p.x, p.Ax)
+    scale!(p.x, one(eltype(p.x)) / norm(p.x))
+
+    p.residual, iteration + 1
 end
 
-#########################
-# Method Implementation #
-#########################
-
-function powm_method!{T}(log::ConvergenceHistory, K::KrylovSubspace{T};
-    tol::Real=eps(real(T))*K.n^3, maxiter::Int=K.n, verbose::Bool=false
-    )
-    verbose && @printf("=== powm ===\n%4s\t%7s\n","iter","resnorm")
-    θ = zero(T)
-    v = Array(T, K.n)
-    for iter=1:maxiter
-        nextiter!(log,mvps=1)
-        v = lastvec(K)
-        y = nextvec(K)
-        θ = dot(v, y)
-        resnorm = real(norm(y - θ*v))
-        push!(log, :resnorm, resnorm)
-        verbose && @printf("%3d\t%1.2e\n",iter,resnorm)
-        resnorm <= tol*abs(θ) && (setconv(log, resnorm >= 0); break)
-        appendunit!(K, y)
-    end
-    verbose && @printf("\n")
-    θ, v
+# Transforms the eigenvalue back whether shifted or inversed
+@inline transform_eigenvalue(θ, inverse::Bool, σ) = σ + (inverse ? inv(θ) : θ)
+
+function powm_iterable!(A, x; tol = eps(real(eltype(A))) * size(A, 2) ^ 3, maxiter = size(A, 1))
+    T = eltype(x)
+    PowerMethodIterable(A, x, tol, maxiter, zero(T), similar(x), similar(x), realmax(real(T)))
+end
+
+function powm_iterable(A; kwargs...)
+    x0 = rand(Complex{real(eltype(A))}, size(A, 1))
+    scale!(x0, one(eltype(A)) / norm(x0))
+    powm_iterable!(A, x0; kwargs...)
 end
 
 ####################
 # API method calls #
 ####################
 
-function invpowm(A;
-    x=nothing, shift::Number=0, tol::Real=eps(real(eltype(A)))*size(A,2)^3,
-    maxiter::Int=size(A,2), log::Bool=false, kwargs...
-    )
-    K = KrylovSubspace(A, 1)
-    x==nothing ? initrand!(K) : init!(K, x/norm(x))
+function powm(A; kwargs...)
```

Comment:
Not related to this PR but we might want to consider renaming this function to something that looks less like a matrix power function.